src/libraries/System.Private.CoreLib/src/System/Guid.cs

```diff
@@ -769,33 +769,33 @@ public bool TryWriteBytes(Span<byte> destination)
         public override int GetHashCode()
         {
             // Simply XOR all the bits of the GUID 32 bits at a time.
-            return _a ^ Unsafe.Add(ref _a, 1) ^ Unsafe.Add(ref _a, 2) ^ Unsafe.Add(ref _a, 3);
+            ref int r = ref Unsafe.AsRef(in _a);
+            return r ^ Unsafe.Add(ref r, 1) ^ Unsafe.Add(ref r, 2) ^ Unsafe.Add(ref r, 3);
         }
 
         // Returns true if and only if the guid represented
         //  by o is the same as this instance.
         public override bool Equals(object? o)
         {
-            Guid g;
-            // Check that o is a Guid first
-            if (o == null || !(o is Guid))
-                return false;
-            else g = (Guid)o;
-
-            // Now compare each of the elements
-            return g._a == _a &&
-                Unsafe.Add(ref g._a, 1) == Unsafe.Add(ref _a, 1) &&
-                Unsafe.Add(ref g._a, 2) == Unsafe.Add(ref _a, 2) &&
-                Unsafe.Add(ref g._a, 3) == Unsafe.Add(ref _a, 3);
+            return o is Guid
+                && EqualsCore(this, Unsafe.Unbox<Guid>(o));
         }
 
-        public bool Equals(Guid g)
+        public bool Equals(Guid g) => EqualsCore(this, g);
+
+        private static bool EqualsCore(in Guid left, in Guid right)
         {
-            // Now compare each of the elements
-            return g._a == _a &&
-                Unsafe.Add(ref g._a, 1) == Unsafe.Add(ref _a, 1) &&
-                Unsafe.Add(ref g._a, 2) == Unsafe.Add(ref _a, 2) &&
-                Unsafe.Add(ref g._a, 3) == Unsafe.Add(ref _a, 3);
+            ref int rA = ref Unsafe.AsRef(in left._a);
+            ref int rB = ref Unsafe.AsRef(in right._a);
+
+            // Compare each element
+
+            if (rA != rB) { return false; }
```

Comment:
Is this code pattern working around JIT limitations?

Reply:
The same reason that `string.IsNullOrEmpty` has that weird _? true : false_ construct at the end. But honestly, since this method isn't getting inlined anywhere, it's really not critical to optimize this. I'll simplify.